acra-censor/handlers/handlers_util.go

```diff
@@ -125,711 +171,1558 @@ func NormalizeAndRedactSQLQuery(sql string) (normalizedQuery string, redactedQue
 	normalizedQ := sqlparser.String(stmt)
 
 	// redact and mask VALUES
-	sqlparser.Normalize(stmt, bv, ValuePlaceholder)
+	sqlparser.Normalize(stmt, bv, ValueMask)
 	redactedQ := sqlparser.String(stmt)
 
 	return normalizedQ, redactedQ, nil
 }
 
-func checkPatternsMatching(patterns [][]sqlparser.SQLNode, query string) (bool, error) {
-	statement, err := sqlparser.Parse(query)
+func checkPatternsMatching(patterns []sqlparser.Statement, query string) (bool, error) {
+	parsedQuery, err := sqlparser.Parse(query)
 	if err != nil {
 		log.WithError(err).Errorln("Can't parse query")
 		return false, ErrQuerySyntaxError
 	}
-	queryNodes, err := getAllNodes(statement)
-	if err != nil {
-		return false, err
-	}
-	for _, singlePatternNodes := range patterns {
-		if checkSinglePatternMatch(queryNodes, singlePatternNodes) {
+
+	for _, pattern := range patterns {
+		if checkSinglePatternMatch(parsedQuery, pattern) {
 			return true, nil
 		}
 	}
 	return false, nil
 }
 
-func checkSinglePatternMatch(queryNodes []sqlparser.SQLNode, patternNodes []sqlparser.SQLNode) bool {
-	matchOccurred := false
-	matchOccurred = handleSelectPattern(queryNodes, patternNodes)
-	if matchOccurred {
+func checkSinglePatternMatch(query, pattern sqlparser.Statement) bool {
+	match := false
+	match = handleUnionStatement(query, pattern)
+	if match {
 		return true
 	}
-	matchOccurred = handleSelectColumnPattern(queryNodes, patternNodes)
-	if matchOccurred {
+	match = handleSelectStatement(query, pattern)
+	if match {
 		return true
 	}
-	matchOccurred = handleSelectWherePattern(queryNodes, patternNodes)
-	if matchOccurred {
+	match = handleStreamStatement(query, pattern)
+	if match {
 		return true
 	}
-	matchOccurred = handleWherePatterns(queryNodes, patternNodes)
-	if matchOccurred {
+	match = handleInsertStatement(query, pattern)
+	if match {
 		return true
 	}
-	matchOccurred = handleStarPattern(queryNodes, patternNodes)
-	if matchOccurred {
+	match = handleUpdateStatement(query, pattern)
+	if match {
 		return true
 	}
-	matchOccurred = handleLimitValuePattern(queryNodes, patternNodes)
-	if matchOccurred {
+	match = handleDeleteStatement(query, pattern)
+	if match {
 		return true
 	}
-	//query doesn't match any stored pattern
-	return false
-}
-
-// handleSelectPattern handles %%SELECT%% pattern
-func handleSelectPattern(queryNodes, patternNodes []sqlparser.SQLNode) bool {
-	if reflect.TypeOf(queryNodes[0]) == reflect.TypeOf(patternNodes[0]) {
-		if patternNodeSelect, ok := patternNodes[0].(*sqlparser.Select); ok && strings.EqualFold(sqlparser.String(patternNodeSelect.SelectExprs), SelectConfigPlaceholderReplacerPart2) {
-			return true
-		}
+	match = handleSetStatement(query, pattern)
```

Comment:
1. the same as above with abstract type for methods, better to use loop with abstract type like `type handleFunction func(query, pattern sqlparser.SQLNode) bool`

2. i think we don't need to add specific function for each type of statement or struct field, only for which we process specially. all other statements and fields should processed with some default logic. if we add method for each statement and process it as 20 other same statements, we add a lot of unnecessary code that we should read and maintain. now we have specific functions only for our custom patterns. for example if i read and see something like `processWhereClause(query, pattern)` - I think that we process it with some specific logic. when i see like `reflect.DeepEqual(query, pattern) - I understand that we doesn't support or haven't special logic for such statements and no matter which values it has

Reply:
I deliberately go out from reflect.DeepEqual - because you said me somehow that it is very cost system call. Currently many false cases are detected by matching types (when types are not equal) - and It seems I used reflect.DeepEqual only two-three times in code. Also such skeleton logic covers all sql cases that sqlparser library can suggest.

Reply:
And also it seems it's very simple to add specific checks, patterns if we will need